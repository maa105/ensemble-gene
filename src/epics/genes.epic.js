import { LOAD_GENE, GENE_LOADED, LOAD_GENE_PROTEINS, LOAD_PROTEIN, GENE_SYMBOL_INPUT_VALUE_CHANGE, HGVS_INPUT_VALUE_CHANGE } from '../actions';
import { noopAction, loadGene, loadingGene, loadGeneProteins, loadingGeneProteins, geneProteinsLoaded, errorLoadingGeneProteins, geneLoaded, errorLoadingGene, loadProtein, loadingProtein, proteinLoaded, errorLoadingProtein } from '../actions';
import { of, concat } from 'rxjs';
import { withLatestFrom, switchMap, catchError, map, debounceTime } from 'rxjs/operators';
import { combineEpics, ofType } from 'redux-observable';
import _ from 'lodash';
import { MIN_GENE_SYMBOL_DIGITS } from '../constants';
import { ajax } from 'rxjs/ajax';

const geneSymbolInputValueChangeEpic = (action$) => action$.pipe(
  ofType(GENE_SYMBOL_INPUT_VALUE_CHANGE),
  debounceTime(1000),
  map((action) => {
    if(action.value.length >= MIN_GENE_SYMBOL_DIGITS) {
      return loadGene(action.value);
    }
    else {
      return noopAction();
    }
  })
);

const hgvsInputValueChangeEpic = (action$, state$) => action$.pipe(
  ofType(HGVS_INPUT_VALUE_CHANGE),
  debounceTime(1000),
  withLatestFrom(state$),
  map(([action,state]) => {
    let proteinId = state.gene.HGVSInputValueValidationResult && state.gene.HGVSInputValueValidationResult.proteinId;
    if(proteinId) {
      return loadProtein(proteinId);
    }
    else {
      return noopAction();
    }
  })
);

const geneProteinsLoadedEpic = action$ => action$.pipe(
  ofType(GENE_LOADED),
  map((action) => loadGeneProteins(action.gene))
);

const loadGeneEpic = (action$, state$, { ajax }) => action$.pipe(
  ofType(LOAD_GENE),
  withLatestFrom(state$),
  switchMap(([action,state]) => {
    const geneSymbol = action.symbol.trim().toLowerCase();
    if(!state.gene.genes[geneSymbol] && 
      !state.gene.loadingGenes[geneSymbol] && 
      (!state.gene.genesWithLoadingError[geneSymbol] || 
        !state.gene.genesWithLoadingError[geneSymbol].notFound)) {
      return concat(of(loadingGene(action.symbol, action.species)),
        ajax.get('http://rest.ensembl.org/lookup/symbol/' + action.species + '/' + geneSymbol.toUpperCase() + '.json?;expand=1',
          { 'Accept': 'application/json' }).pipe(
          map(result => {
            return geneLoaded(result.response, action.symbol, action.species);
          }),
          catchError((err) => of(errorLoadingGene(err, action.symbol, action.species)))
        )
      );
    }
    else {
      return of(noopAction());
    }
  })
);

const loadGeneProteinsEpic = (action$, state$, { ajax }) => action$.pipe(
  ofType(LOAD_GENE_PROTEINS),
  withLatestFrom(state$),
  switchMap(([action, state]) => {
    const geneSymbol = action.gene.display_name.toLowerCase();
    if(!state.gene.proteinsByGene[geneSymbol] && 
      !state.gene.loadingGeneProteins[geneSymbol] && 
      (!state.gene.genesWithLoadingError[geneSymbol] ||
        !state.gene.genesWithLoadingError[geneSymbol].notFound)) {
      const ids =_.map(action.gene.Transcript, 'id');
      return concat(of(loadingGeneProteins(action.gene)),
        ajax.post('http://rest.ensembl.org/sequence/id', 
          '{"ids":' + JSON.stringify(ids) + ',"type":"protein"}',
          { 'Content-type': 'application/json', 'Accept': 'application/json' }).pipe(
          map(result => {
            return geneProteinsLoaded(result.response, action.gene);
          }),
          catchError((err) => of(errorLoadingGeneProteins(err, action.gene)))
        )
      );
    }
    else {
      return of(noopAction());
    }
  })
);

const loadProteinEpic = (action$, state$, { ajax }) => action$.pipe(
  ofType(LOAD_PROTEIN),
  withLatestFrom(state$),
  switchMap(([action, state]) => {
    const proteinId = action.proteinId.trim().toLowerCase();
    if(!state.gene.proteins[proteinId] && 
      !state.gene.loadingProteins[proteinId] && 
      (!state.gene.proteinsWithLoadingError[proteinId] ||
        !state.gene.proteinsWithLoadingError[proteinId].notFound)) {
      return concat(of(loadingProtein(proteinId)),
        ajax.get('http://rest.ensembl.org/sequence/id/' + proteinId + '?content-type=application/json', 
          { 'Content-type': 'application/json', 'Accept': 'application/json' }).pipe(
          map(result => {
            return proteinLoaded(result.response, proteinId);
          }),
          catchError((err) => of(errorLoadingProtein(err, proteinId)))
        )
      );
    }
    else {
      return of(noopAction());
    }
  })
);

const genesEpic = (...args) => combineEpics(
  geneSymbolInputValueChangeEpic,
  hgvsInputValueChangeEpic,
  geneProteinsLoadedEpic,
  loadGeneEpic,
  loadGeneProteinsEpic,
  loadProteinEpic
)(...args, { ajax });

export default genesEpic;
