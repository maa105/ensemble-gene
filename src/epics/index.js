import { combineEpics } from 'redux-observable';
import genesEpic from './genes.epic';

export const rootEpic = combineEpics(
  genesEpic
);
