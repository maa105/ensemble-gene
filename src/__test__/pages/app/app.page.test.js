import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import App from '../../../pages/app/app.page';
import { createEpicMiddleware } from 'redux-observable';
import { applyMiddleware, createStore } from 'redux';
import { rootEpic } from '../../../epics';
import rootReducer from '../../../reducers';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {

  const epicMiddleware = createEpicMiddleware();
  const store = createStore(rootReducer, applyMiddleware(epicMiddleware));
  epicMiddleware.run(rootEpic);
  
  const div = document.createElement('div');
  ReactDOM.render(<Router>
    <Provider store={store}>
      <App />
    </Provider>
  </Router>, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('app should render correctly', () => {

  const epicMiddleware = createEpicMiddleware();
  const store = createStore(rootReducer, applyMiddleware(epicMiddleware));
  epicMiddleware.run(rootEpic);

  const component = renderer.create(
    <Router>
      <Provider store={store}>
        <App />
      </Provider>
    </Router>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
