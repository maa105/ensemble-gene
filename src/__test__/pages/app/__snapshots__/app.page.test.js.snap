// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app should render correctly 1`] = `
<div>
  <nav
    className="navbar navbar-inverse"
  >
    <div
      className="container"
    >
      <div
        className="navbar-header"
      >
        <a
          className="navbar-brand"
          href="https://www.ensembl.org"
        >
          Ensembl
        </a>
        <button
          className="navbar-toggle collapsed"
          onClick={[Function]}
          type="button"
        >
          <span
            className="sr-only"
          >
            Toggle navigation
          </span>
          <span
            className="icon-bar"
          />
          <span
            className="icon-bar"
          />
          <span
            className="icon-bar"
          />
        </button>
      </div>
      <div
        aria-expanded={null}
        className="navbar-collapse collapse"
      >
        <ul
          className="nav navbar-nav"
          role={null}
        >
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="/find-proteins"
              onClick={[Function]}
              onKeyDown={[Function]}
            >
              Find Protein
            </a>
          </li>
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="/hgvs"
              onClick={[Function]}
              onKeyDown={[Function]}
            >
              HGVS
            </a>
          </li>
        </ul>
        <ul
          className="nav navbar-nav navbar-right"
          role={null}
        >
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="https://rest.ensembl.org/"
              onClick={[Function]}
              onKeyDown={[Function]}
            >
              API
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <div
      className="form-group has-feedback"
    >
      <label
        className="control-label"
        htmlFor="geneSymbol"
      >
        Gene Symbol
      </label>
      <input
        className="form-control"
        id="geneSymbol"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="e.g. BRAF"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group has-feedback"
    >
      <label
        className="control-label"
        htmlFor="proteinIndex"
      >
        Protien Index
      </label>
      <input
        className="form-control"
        id="proteinIndex"
        onChange={[Function]}
        placeholder="e.g. 600"
        type="number"
        value={600}
      />
    </div>
    <div
      className="form-group has-feedback"
    >
      <label
        className="control-label"
        htmlFor="proteinValue"
      >
        Protien Value
      </label>
      <select
        className="form-control"
        id="proteinValue"
        onChange={[Function]}
        placeholder="Protein letter e.g. V"
        type="text"
        value="V"
      >
        <option
          value="A"
        >
          A
           - 
          Alanine
           (
          Ala
          )
        </option>
        <option
          value="C"
        >
          C
           - 
          Cysteine
           (
          Cys
          )
        </option>
        <option
          value="D"
        >
          D
           - 
          Aspartic Acid
           (
          Asp
          )
        </option>
        <option
          value="E"
        >
          E
           - 
          Glutamic Acid
           (
          Glu
          )
        </option>
        <option
          value="F"
        >
          F
           - 
          Phenylalanine
           (
          Phe
          )
        </option>
        <option
          value="G"
        >
          G
           - 
          Glycine
           (
          Gly
          )
        </option>
        <option
          value="H"
        >
          H
           - 
          Histidine
           (
          His
          )
        </option>
        <option
          value="I"
        >
          I
           - 
          Isoleucine
           (
          Ile
          )
        </option>
        <option
          value="K"
        >
          K
           - 
          Lysine
           (
          Lys
          )
        </option>
        <option
          value="L"
        >
          L
           - 
          Leucine
           (
          Leu
          )
        </option>
        <option
          value="M"
        >
          M
           - 
          Methionine
           (
          Met
          )
        </option>
        <option
          value="N"
        >
          N
           - 
          Asparagine
           (
          Asn
          )
        </option>
        <option
          value="P"
        >
          P
           - 
          Proline
           (
          Pro
          )
        </option>
        <option
          value="Q"
        >
          Q
           - 
          Glutamine
           (
          Gln
          )
        </option>
        <option
          value="R"
        >
          R
           - 
          Arginine
           (
          Arg
          )
        </option>
        <option
          value="S"
        >
          S
           - 
          Serine
           (
          Ser
          )
        </option>
        <option
          value="T"
        >
          T
           - 
          Threonine
           (
          Thr
          )
        </option>
        <option
          value="V"
        >
          V
           - 
          Valine
           (
          Val
          )
        </option>
        <option
          value="W"
        >
          W
           - 
          Tryptophan
           (
          Trp
          )
        </option>
        <option
          value="X"
        >
          X
           - 
          X
           (
          X
          )
        </option>
        <option
          value="Y"
        >
          Y
           - 
          Tyrosine
           (
          Tyr
          )
        </option>
      </select>
    </div>
  </div>
</div>
`;
